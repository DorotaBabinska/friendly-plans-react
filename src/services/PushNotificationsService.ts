import { Platform } from 'react-native';
import firebase, { RNFirebase } from 'react-native-firebase';

import { i18n } from 'locale';

export default class PushNotificationsService {
  initialize = async () => {
    const hasPermission = await this.checkPermission();
    /* istanbul ignore else */
    if (hasPermission) {
      this.createNotificationChannel();
      this.createNotificationListeners();
    }
  };

  checkPermission = async (): Promise<boolean> => {
    const enabled = await firebase.messaging().hasPermission();
    /* istanbul ignore else */
    if (enabled) {
      const token = await this.getToken();
      return !!token;
    } else {
      return this.requestPermission();
    }
  };

  getToken = (): Promise<string> => {
    return firebase.messaging().getToken();
  };

  requestPermission = async (): Promise<boolean> => {
    try {
      await firebase.messaging().requestPermission();
      const token = await this.getToken();
      return !!token;
    } catch (error) {
      /* User has rejected permissions */
      /* istanbul ignore next */
      return false;
    }
  };

  createNotificationChannel = () => {
    /* istanbul ignore next */
    if (Platform.OS === 'android') {
      const channel = new firebase.notifications.Android.Channel(
        'Notifications',
        i18n.t('notifications:channelName'),
        firebase.notifications.Android.Importance.Max,
      ).setDescription(i18n.t('notifications:channelDescription'));
      firebase.notifications().android.createChannel(channel);
    }
  };

  createNotificationListeners = async () => {
    await firebase.messaging().getToken();
    firebase.notifications().onNotification(this.handleNotification);
    firebase.notifications().onNotificationOpened(this.handleNotificationOpen);

    const notificationOpen = await firebase
      .notifications()
      .getInitialNotification();

    /* istanbul ignore else */
    if (notificationOpen) {
      this.handleNotificationOpen(notificationOpen);
    }
  };

  /* Handle notifications within the app when app is running in foreground */
  /* istanbul ignore next */
  handleNotification = (notification: RNFirebase.notifications.Notification) =>
    null;

  /* Handle notifications when app is open from the notification */
  /* istanbul ignore next */
  handleNotificationOpen = (
    notification: RNFirebase.notifications.NotificationOpen,
  ) => null;
}
