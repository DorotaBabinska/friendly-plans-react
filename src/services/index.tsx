import _AnalyticsService from './AnalyticsService';
import _CrashlyticsService from './CrashlyticsService';
import _NavigationService from './NavigationService';
import _PushNotificationsService from './PushNotificationsService';

export const NavigationService = new _NavigationService();
export const PushNotificationsService = new _PushNotificationsService();
export const AnalyticsService = new _AnalyticsService(NavigationService);
export const CrashlyticsService = new _CrashlyticsService();
