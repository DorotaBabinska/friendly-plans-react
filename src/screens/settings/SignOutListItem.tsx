import React from 'react';
import firebase from 'react-native-firebase';

import { ListItem } from 'components';
import { i18n } from 'locale';
import { NavigationService } from 'services';

export class SignOutListItem extends React.PureComponent {
  signOut = async () => {
    await firebase.auth().signOut();
    NavigationService.navigate('Unauthenticated');
  };

  openSignOutDialog = () => {
    NavigationService.navigate('Dialog', {
      title: i18n.t('settings:signOutTitle'),
      description: i18n.t('settings:signOutSubtitle'),
      onPress: this.signOut,
      buttonTitle: i18n.t('settings:signOutAction'),
    });
  };

  render() {
    return (
      <ListItem
        icon={{ name: 'logout-variant' }}
        title={i18n.t('settings:signOutTitle')}
        subtitle={i18n.t('settings:signOutSubtitle')}
        onPress={this.openSignOutDialog}
      />
    );
  }
}
