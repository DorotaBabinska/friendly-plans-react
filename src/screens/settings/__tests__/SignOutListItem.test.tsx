import { shallow } from 'enzyme';
import React from 'react';

import { NavigationService } from 'services';
import { SignOutListItem } from '../SignOutListItem';

describe('SignOutListItem', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<SignOutListItem />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should open confirmation dialog on press', () => {
    const wrapper = shallow(<SignOutListItem />);
    const spy = jest
      .spyOn(NavigationService, 'navigate')
      .mockImplementation(jest.fn());
    wrapper.simulate('press');
    expect(spy).toBeCalledWith('Dialog', expect.objectContaining({}));
  });

  it('should navigate to unauthenticated view after logout', async () => {
    const wrapper = shallow(<SignOutListItem />);
    const instance = wrapper.instance() as SignOutListItem;
    const spy = jest
      .spyOn(NavigationService, 'navigate')
      .mockImplementation(jest.fn());
    await instance.signOut();
    expect(spy).toBeCalledWith('Unauthenticated');
  });
});
