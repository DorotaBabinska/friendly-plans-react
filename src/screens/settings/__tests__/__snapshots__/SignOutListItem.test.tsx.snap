// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignOutListItem should match snapshot 1`] = `
<ListItem
  icon={
    Object {
      "name": "logout-variant",
    }
  }
  onPress={[Function]}
  subtitle="After you sign out, you will need to sign in again."
  title="Sign out"
/>
`;
